google-cloud-storage/src/test/java/com/google/cloud/storage/spi/v1/HttpStorageRpcTest.java

```diff
@@ -0,0 +1,174 @@
+import com.google.api.core.ApiClock;
+import com.google.api.services.storage.model.Bucket;
+import com.google.cloud.ServiceOptions;
+import com.google.cloud.http.HttpTransportOptions;
+import com.google.cloud.storage.StorageException;
+import com.google.cloud.storage.StorageOptions;
+import com.google.cloud.storage.spi.StorageRpcFactory;
+import com.google.common.io.ByteStreams;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+import java.util.zip.GZIPInputStream;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+public class HttpStorageRpcTest {
+
+  // object under test, created before each test case
+  private HttpStorageRpc rpc;
+
+  // Objects for a test case: one object per one RPC mock.
+  // A test creates as many MockData objects as it issues RPC calls.
+  private Queue<MockData> testMockData;
+
+  private static final String BASE_URL = "https://storage.googleapis.com/storage/v1/b";
+  private static final String URL_PROJECT = "project=projectId&projection=full";
+
+  private static final Answer<Object> UNEXPECTED_CALL_ANSWER =
+      new Answer<Object>() {
+        @Override
+        public Object answer(InvocationOnMock invocation) {
+          throw new IllegalArgumentException(
+              "Unexpected call of "
+                  + invocation.getMethod()
+                  + " with "
+                  + Arrays.toString(invocation.getArguments()));
+        }
+      };
+
+  private static final ApiClock TIME_SOURCE =
+      new ApiClock() {
+        @Override
+        public long nanoTime() {
+          return 42_000_000_000L;
+        }
+
+        @Override
+        public long millisTime() {
+          return 42_000L;
+        }
+      };
+
+  private static final StorageRpcFactory RPC_FACTORY_MOCK =
+      mock(StorageRpcFactory.class, UNEXPECTED_CALL_ANSWER);
+
+  /**
+   * Adds mock parameters for the RPC call.
+   *
+   * @param method expected HTTP method
+   * @param url expected URL
+   * @param response HTTP response object to be returned
+   * @return a holder object which will contain the Http Request made by RPC
+   */
+  private RpcRequestHolder mockResponse(String method, String url, LowLevelHttpResponse response) {
+    RpcRequestHolder holder = new RpcRequestHolder();
+    testMockData.add(new MockData(method, url, response, holder));
+    return holder;
+  }
+
+  private static String getUnzippedContent(LowLevelHttpRequest request) throws IOException {
+    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+    request.getStreamingContent().writeTo(outputStream);
+    GZIPInputStream gzipInputStream =
+        new GZIPInputStream(new ByteArrayInputStream(outputStream.toByteArray()));
```

Comment:
since you already have a byte array use an inflater, not a stream

Reply:
Would you mind if I kept using GZIPInputStream? java.util.zip.Inflater introduces new DataFormatException the test should handle. The inflater requires a byte buffer of a certain size, but the size is unknown. And I think, I can't use the java.util.zip.Inflater from the box to decompress GZIP content: 
```
 To fully decode the gzip format with this class, you will need to process the gzip header as described in RFC 1952, use inflater to decompress the raw deflate data, compute the crc32 of the uncompressed data using that class, and then verify the crc and length (modulo 2^32) in the gzip trailer, again as specified in the RFC.
```

Or do you mean another inflater?

Reply:
Hmm, it's weird to copy this into a  byte array first. You should be able to go straight from the LowLevelHttpRequest to the decompressor, but LowLevelHttpRequest doesn't offer an input stream top read from, which is really strange. Maybe use a piped stream but that would need an extra thread.